ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS/DNSActions.hs

```diff
@@ -234,45 +238,49 @@ asyncResolverResource resolvConf =
       _ -> do
         constantResource <$> getResolver resolvConf
   where
-    handlers :: FilePath -> StrictTVar IO TimedResolver
-             -> [Handler IO
-                  ( Either (DNSorIOError IOException) DNS.Resolver
-                  , Resource IO (DNSorIOError IOException) DNS.Resolver)]
-    handlers filePath resourceVar =
-      [ Handler $
-          \(err :: IOException) ->
-            pure (Left (IOError err), go filePath resourceVar)
-      , Handler $
-          \(err :: DNS.DNSError) ->
-            pure (Left (DNSError err), go filePath resourceVar)
-      ]
+    handlers :: [ Handler IO (Either (DNSorIOError IOException) a) ]
+    handlers  = [ Handler $ pure . Left . IOError
+                , Handler $ pure . Left . DNSError
+                ]
 
     go :: FilePath -> StrictTVar IO TimedResolver
        -> Resource IO (DNSorIOError IOException) DNS.Resolver
     go filePath resourceVar = Resource $ do
-      r <- atomically (readTVar resourceVar)
+      r <- readTVarIO resourceVar
       case r of
         NoResolver ->
           do
-            modTime <- getModificationTime filePath
-            resolver <- getResolver resolvConf
-            atomically (writeTVar resourceVar (TimedResolver resolver modTime))
-            pure (Right resolver, go filePath resourceVar)
-          `catches` handlers filePath resourceVar
-
-        TimedResolver resolver modTime ->
-          do
-            modTime' <- getModificationTime filePath
+            result
+              <- (curry Right
+                   <$> getTimeStamp filePath
+                   <*> getResolver resolvConf)
+                 `catches` handlers
+            case result of
+              Left err ->
+                pure (Left err, go filePath resourceVar)
+              Right (modTime, resolver) -> do
+                atomically (writeTVar resourceVar (TimedResolver resolver modTime))
+                pure (Right resolver, go filePath resourceVar)
+
+        TimedResolver resolver modTime -> do
+          result <- runExceptT $ do
+            modTime' <- ExceptT $ (Right <$> getTimeStamp filePath)
+                                  `catches` handlers
             if modTime' <= modTime
-                then pure (Right resolver, go filePath resourceVar)
-                else do
-                  resolver' <- getResolver resolvConf
-                  atomically (writeTVar resourceVar (TimedResolver resolver' modTime'))
-                  pure (Right resolver', go filePath resourceVar)
-          `catches` handlers filePath resourceVar
-#else
-asyncResolverResource resolvConf = resolverResource resolvConf
-#endif
+              then return resolver
+              else do
+              resolver' <- ExceptT $ (Right <$> getResolver resolvConf)
+                                     `catches` handlers
+              lift $ atomically (writeTVar resourceVar (TimedResolver resolver' modTime'))
```

Comment:
Is this lift necessary? Can't we perform this outside `ExceptT` in case of success (`Right`) ? it could avoid one extra dependency

Reply:
I thought we don't have `MonadSTM` instance for `ErrorT`, but we actually have [one](https://github.com/input-output-hk/io-sim/blob/main/io-classes/src/Control/Monad/Class/MonadSTM.hs#L1049), maybe I need to update `io-sim` hash.

I think we cannot take it out without the if condition, and that's probably not worth it.

Reply:
we can remove `lift` but we'll need [`castStrictTVar`](https://github.com/input-output-hk/io-sim/blob/main/strict-stm/src/Control/Monad/Class/MonadSTM/Strict.hs?plain=1#L120-L123).